### ISSUE TITLE
viewmodels - modules with dependencies

### ISSUE THREAD

**Issue by johannes-z** [vuejs/Discussion#203]
I read the "Building Larger Apps" and "Tips & Best Practices" sections on vuejs.org, but I still have some questions regarding this topic. The problem with the documentation is, that it's just very basic, and doesn't use any other libraries or self-made javascript code.
As far as I understood the documentation, modularization is (only) easily done via the npm vue.js module - and I agree, it's very easy to set up a VM, and require all the dependencies. The problem I have though, is that I have to use the code of other libraries, which may or may not be a npm module themselves (for example window hooked objects). 

What I have right now are .html templates (for the blade template engine), and at the bottom of each such file, I have a script block, defining my viewmodel. Sometimes more than 1 viewmodel, and since they can be more or less complex, the script section gets very bloated, and the code isn't reuseable.

A reallife example from my application would be:
```html
<script lang="text/javascript">
        (function () {
            var NetMsg = App.NetMsg;
            var client = new App.Client();

            App.registerViewModel({
                el: '#messages',
                data: {...},
                ready: function () {
                    client.sendData(new NetMsg(NetMsg.verb.GET, NetMsg.action.MESSAGES))
                },
                ...
            });
        })();
</script>
```
App.registerViewModel is a very basic function, that binds the viewmodel to the "App" namespace by the "el" selector - making sure, only one such viewmodel exists in the application. This is needed because I use websockets, and have to handle the data in the "onMessage" event of the connection.
There's also a method in the viewmodel, which uses the client variable. 
What I really want to do is something like this in my script tag in each such template file:
```html
<script lang="text/javascript">
	(function () {
		var NetMsg = App.NetMsg;
		var client = new App.Client();

		vm_Messages(client);
	})();
</script>
```
whereas vm_Messages() does the same as the inline code, but with it's components moved into their own files.

**TL;DR:** Modularize viewmodels (with npm, or anything else) but still being able to use "custom code" in the viewmodels methods, components, etc. Making the code more readable.

I don't know if this is approach is ok or not, but it's the only thing that came into my mind. I'd really appreciate your thoughts on the matter.

**Comment by yyx990803** [comment]
You can of course use globals exposed on `window` - webpack/browserify bundles eventually run in the browser.

**Comment by johannes-z** [comment]
The problem I have is exposing the defined viewmodels with its resolved dependencies in the ```window```. I don't rather, as described, call a function that creates the Vue instance, than create it in an inline script tag. Mainly, because I want to use the proposed modulization of vue viewmodels.

**Comment by yyx990803** [comment]
you can also expose things on `window` inside the modules, e.g. `window.myComponent = { ... }`, and then in your bundle entry file make sure to just `require()` them.

**Comment by johannes-z** [comment]
welp, didn't think of that. Following your suggestion, binding the functions directly to the window in the module let's me access them. Thanks a lot!
